include/vcpkg/base/stringview.h

```diff
@@ -39,23 +33,56 @@ namespace vcpkg
         constexpr size_t size() const noexcept { return m_size; }
         constexpr bool empty() const noexcept { return m_size == 0; }
 
+        // intentionally not provided because this may not be null terminated
+        // constexpr const char* c_str() const
+
         std::string to_string() const;
         void to_string(std::string& out) const;
         explicit operator std::string() const { return to_string(); }
 
         StringView substr(size_t pos, size_t count = std::numeric_limits<size_t>::max()) const noexcept;
 
         constexpr char operator[](size_t pos) const noexcept { return m_ptr[pos]; }
+        friend std::string operator+(std::string&& l, const StringView& r);
 
     private:
         const char* m_ptr = 0;
         size_t m_size = 0;
     };
 
+    // Intentionally not hidden friends to allow comparison of Path, for example
     bool operator==(StringView lhs, StringView rhs) noexcept;
     bool operator!=(StringView lhs, StringView rhs) noexcept;
     bool operator<(StringView lhs, StringView rhs) noexcept;
     bool operator>(StringView lhs, StringView rhs) noexcept;
     bool operator<=(StringView lhs, StringView rhs) noexcept;
     bool operator>=(StringView lhs, StringView rhs) noexcept;
+
+    // A counted view of a null-terminated string
+    struct ZStringView : StringView
+    {
+        constexpr ZStringView() : StringView("", size_t{}) { }
+        ZStringView(const std::string& s) : StringView(s) { }
+        constexpr ZStringView(const char* ptr, size_t size) noexcept : StringView(ptr, size) { }
+        // intentionally not provided to discourage non-null-termination:
+        // constexpr ZStringView(const char* b, const char* e)
+        ZStringView(const char* ptr) noexcept : StringView(ptr) { }
+
+        constexpr const char* c_str() const noexcept { return data(); }
+
+        // Note that only the 1 parameter version of substr is provided to preserve null termination
+        // (The name from the base class is intentionally hidden)
+        ZStringView substr(size_t pos) const noexcept;
```

Comment:
```suggestion
        // Note that when `substr` only takes one parameter, it can return a `ZStringView`; otherwise it must return a `StringView`.
        ZStringView substr(size_t pos) const noexcept;
        StringView substr(size_t pos, size_t count) const noexcept
        {
             return StringView::substr(pos, count);
         }
```

`zstr.substr(1, 3)` is fine, it just needs to return a `StringView`.

Reply:
I considered doing this, but since I couldn't come up with any actual users of such a thing I did not add it.